ecosystem/sep-0001.md

```diff
@@ -116,9 +116,10 @@ These fields go in the `stellar.toml` `[[CURRENCIES]]` list, one set of fields f
 
 Field | Requirements | Description
 ------|--------------|------------
-code | string (<= 12 char) | Token code
+code | string (<= 12 char) | Asset code. Required for assets that are issued using a Stellar account as opposed to a contract.
+issuer | `G...` string | Asset issuer Stellar public key. Required for assets that are issued using a Stellar account as opposed to a contract.
+contract | `C...` string | Contract ID of the asset contract. The asset must be compatible with the [SEP-41 Token Interface](sep-0041.md) to be defined here.
```

Comment:
Is the contract field optional? We should specify that.

Reply:
Well, contracts for Stellar assets are technically optional, since it requires someone to deploy the SAC for it. At the same time, it should be listed if a contract for the asset exists. I'm not sure how to phrase this.

Reply:
Anyone can create the contract for a stellar asset, it doesn't have to be the issuer, so I'm not sure there is a natural dependency here that if one exists the other should exist.

The contract is always derivable from the code and issuer and network passphrase, so maybe we should never specify the contract for stellar assets and make issuer and contract mutually exclusive. That way applications don't make the assumption the field will be present.

I think it's preferred to assume a contract always exists, and applications that interact with the contract need to be resilient to it not being created yet and creating it if it doesn't exist, or restoring it if it has been archived due to a lack of use.

Reply:
While it makes things harder for clients, I think making (code, issuer) and (contract) mutually exclusive makes sense. This makes it clear whether there is a Stellar asset (and presumably a contract) or just a contract.

It also could be more secure or less error prone. What are the consequences of a client matching a currency listed in a stellar.toml by its (code, issuer) and then interacting with a malicious contract listed alongside the other fields?

Reply:
Agreed it could be more secure. If the contract is provided for stellar assets then it will be possible for applications to depend on it being correct, and a compromised .toml file could change the contract ID while keeping the issuer and code the same, causing applications to interact with a contract for a different issuer. Leaving contract out forces everyone to act the same way with a single set of inputs.

Reply:
Orbit was in favor of always including the contract so I'd like to wait to get his input, but if we all agree I'll update the PR.

Reply:
The contract always being included would be convenient for applications that will just trust that it is correct, but it also sets them up to be compromised, and they really should be validating it is correct so in either case running through the same logic of generating the contract ID then checking it matches.

Reply:
cc @orbitlens Thoughts?

Reply:
I'm opposed to automatically deriving `contract` from the `code`+`issuer` combination for Classic assets.
Reasons:
- The client (e.g. wallet) will **always** has to do additional unnecessary computations to derive the contract address, which might be expensive in terms of performance for cases when a client needs to process metadata for large number of assets at a time (for example, render account balances or show transactions history). Just to understand, sometimes we have to pull metadata for **hundreds** of assets from different issuers just to render a single page.
- Automatic contract address derivation doesn't necessarily mean that the contract actually **exists** on the ledger. If nobody deployed an SAC contract, then wallet needs to verify the SAC contract existence before executing some operations (this won't affect most contract interactions that should be covered by the preflight simulation request, but might pose unnecessary problems for pulling SAC contract statistics and balances).
- Some issuers might want to implement custom logic for their Classic assets on the Soroban side, providing custom token implementation instead of the built-in SAC.

@leighmcculloch 's argument regarding additional risks posed by a compromised TOML file is totally valid, but I don't think we should prevent issuers from creating custom contracts for their Classic assets. We don't discard SEP-24 and SEP-31 only because they both are potentially vulnerable to the same attack.

Reply:
Thanks for the insights. I had been thinking that if a contract address was different to the contract for the code and issuer, that that'd be either a mistake, or malicious event. I hadn't considered that it could be valid for an issuer to want to represent their one conceptual token with two tokens on Stellar, the classic one and the Soroban one in a custom contract.

It's worth noting that in that case where an issuer has a stellar asset and a custom contract, the stellar asset still has a stellar asset contract that anyone can deploy, and I anticipate there will be tooling that automatically derives the contract ID, especially for decentralized tokens that have no toml file, but also in tools that don't wish to talk to the outside world to determine what the contract ID is.

Reply:
> Some issuers might want to implement custom logic for their Classic assets on the Soroban side, providing custom token implementation instead of the built-in SAC.

> I hadn't considered that it could be valid for an issuer to want to represent their one conceptual token with two tokens on Stellar, the classic one and the Soroban one in a custom contract.

I do not think it is valid to represent two completely distinct tokens as one, and I can't think about why anyone would want to do this. If you want custom logic for your asset, in what scenario does it still make sense to have a Stellar asset that is presented to the user as the same token? These two tokens would have different functionality and they would be held in two separate balance entries (one trustline & one contract data). Unifying the two under one UX would almost certainly result is weird user flows, since only a portion of the total unified balance would support whatever custom functionality was implemented in the contract.

Reply:
I updated the PR to have `issuer` and `contract` fields be mutually exclusive. I figured `code` should be present for all entries.